Use `hvPlot` in Plotting API, Add Polygon Projection Support

philipc2 opened the pull request:
<!--  The PR title should summarize the changes, for example "Add `Grid._build_face_dimension` function".
      Avoid non-descriptive titles such as "Addresses issue #229". -->

<!--  Replace XXX with the number of the issue that this PR will resolve. If this PR closed more than one,
      you may add a comma separated sequence-->
Closes #849 #493 #918 #625 

## Overview
<!--  Please provide a few bullet points summarizing the changes in this PR. This should include
      points on any bug fixes, new functions, or other changes that have been made. -->
* Updates plotting API to use `hvPlot`
* Adds support for Cartopy Projections when plotting polygons (raster & vector)
* Updated  `to_geoataframe()` methods to better align with recent changes to `to_polycollection()`
* Adds support for directly constructing a `geopandas.GeoDataFrame` using the `engine='geopandas'` parameter

aaronzedwick commented:
Is there a better mesh you could use for the notebook example? Maybe with more cells? And possibly without the holes? It can be confusing to look at. Also now that I am thinking about it, we should start using the UXarray colormaps for the notebooks.

philipc2 commented:
> Is there a better mesh you could use for the notebook example? Maybe with more cells? And possibly without the holes? It can be confusing to look at. Also now that I am thinking about it, we should start using the UXarray colormaps for the notebooks.

Agree with all the points here. I thought the PR was ready for review (but quickly turned it back to draft). I'll address this points!

erogluorhan commented on uxarray/plot/accessor.py:
```diff
@@ -27,232 +34,195 @@ def __init__(self, uxgrid: Grid) -> None:
-        return grid_plot.node_coords(
-            self._uxgrid, backend=backend, width=width, height=height, **kwargs
-        )
+        return points_df.hvplot.points("lon", "lat", **kwargs)
```
I am a bit confused, starting with this function in `accessor.py`, it does not any longer act like accessor, the functions added with this PR seem to have their own implementations rather than behaving as accessors to the actual functions of `dataarray_plot.py` and `grid_plot.py`. 

Could you please clarify?

philipc2 replied:
I don't believe `dataarray_plot` and `grid_plot` will be necessary moving forward with these changes. The previous separation was consistent with `xarray`, however now that we are much more elegantly wrapping `hvPlot`, we don't need to have it split up as much as we did before. 

It still acts as an accessor (i.e. `GridPlotAccessor` and `UxDataArrayPlotAccessor` are the actual accessor classes) that are attached to the `.plot` attribute of their respective objects.

erogluorhan replied:
Thanks for clarifying! However, then it is still confusing to see we still keep functions in `grid_plot.py` and `dataarray_plot.py` that will no longer be used, e.g. grid_plot.mesh() etc. 

I get that in a future PR (likely when we are addressing the deprecation of functions like `rasterize`), we will remove `grid_plot.py` and `dataarray_plot.py` fully, but even for the meantime, we need to get rid of those functions they are now implemented in `accessor.py` and will no longer be called from either of those "_plot.py" scripts.

Does that make sense?

philipc2 replied:
This makes complete sense, I forgot to remove the functions that aren't called anymore.
* Nothing is being called from `grid_plot.py`, so I have removed that module.
* `dataarray_plot.py` has been reduced to only include the functions that are planned to be deprecated. 

Thanks for clarifying!

erogluorhan replied:
Sounds good!

erogluorhan reviewed (commented):
Please address the comments below

philipc2 commented on benchmarks/mpas_ocean.py:
```diff
@@ -81,8 +81,6 @@ class GeoDataFrame(DatasetBenchmark):
         self.uxds[data_var].to_geodataframe(exclude_antimeridian=exclude_antimeridian)
 
-    def peakmem_to_geodataframe(self, resolution, exclude_antimeridian):
-        gdf = self.uxds[data_var].to_geodataframe(exclude_antimeridian=exclude_antimeridian)
```
> Note
> 
> The peak memory benchmark also counts memory usage during the setup routine, which may confound the benchmark results. One way to avoid this is to use setup_cache instead.

From the ASV documentation. I can include it back, however I chose to remove it since we aren't correctly implementing it.

erogluorhan replied:
Thanks for clarifying! No need to add it back

philipc2 commented on test/test_projection.py:
```diff
@@ -0,0 +1,16 @@
+def test_geodataframe_projection():
+    uxgrid = ux.open_grid(gridfile_geos_cs)
+
+    gdf = uxgrid.to_geodataframe(projection=ccrs.Robinson(), periodic_elements='exclude')
```
Yeah, we might need to reconsider and standardize our approach to testing, maybe having some better parameterized tests and cleaner organizations. 

Think it was around the time of the refactor that the testing started to become very convoluted.

erogluorhan replied:
Ok, opened #986 to follow up on this after this PR.

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1544,51 +1575,149 @@ def to_xarray(self, grid_format: Optional[str] = "ugrid"):
+        """Sets the central longitude of a grid to a desired value.
+
+        Used for visualization of projections with different central
+        longitude values.
```
Keeping under the `Grid` is consistent of operations that directly modify the `Grid`. Unless you were thinking of making this internal and having the user not have control over this?

philipc2 replied:
@erogluorhan 

I am moving this out of the `Grid` and going to implement this internally within plotting.

erogluorhan replied:
Ok

philipc2 commented on uxarray/grid/geometry.py:
```diff
@@ -151,80 +157,147 @@ def _grid_to_polygon_geodataframe(grid, exclude_antimeridian):
+def _build_geodataframe_with_antimeridian(
+    polygon_shells,
+    projected_polygon_shells,
+    projection,
```
Since this is an internal helper, it always is provided a default value through the user-api calls.

erogluorhan replied:
Sounds good

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -162,16 +164,25 @@ def __init__(
         self._gdf = None
         self._gdf_exclude_am = None
+        self._gdf_periodic_state = None
+        self._gdf_projection_state = None
```
I'm using a dictionary now instead of hard-coded parameters. Let me know what you think.

erogluorhan replied:
Definitely looks better

rajeeja commented:
Can you elaborate more on why specifically we picked hvplot? Did we consider others?

philipc2 commented:
> Can you elaborate more on why specifically we picked hvplot? Did we consider others?

<img width="1125" alt="image" src="https://github.com/user-attachments/assets/472d6dfa-25a6-4604-86c5-abac38d779eb">

We were previously directly interfacing with HoloViews, Datashader, and GeoViews. hvPlot provides a clean wrapper to these libraries which means that our plotting implementation is significantly more light weight, while maintaining the same level of functionality. 

hvPlot also has great GeoViews support, meaning that polygon projections can be done efficiently.

rajeeja reviewed (approved):
Both Spatial Pandas and GeoPandas offer a GeoDataFrame, and we use both those libraries and functionalities in our package now!! We use the former for visualization and the latter for its rich IO support.

At this point they sort of live in their separate worlds - with no name mangling, but it could be a little confusing as we build more functionality. 

That aside, hvplot is a good addition, it brings about more flexibility and I'm hoping faster speeds also for visualization.

philipc2 commented:
@erogluorhan 

hvPlot just came out with a new [release](https://github.com/holoviz/hvplot/releases/tag/v0.11.0), which includes `pixel_ratio` support. I've removed the fallback to our old implementation if pixel_ratio was provided as a parameter.

erogluorhan commented on uxarray/grid/grid.py:
```diff
@@ -1544,51 +1565,147 @@ def to_xarray(self, grid_format: Optional[str] = "ugrid"):
+
+        if periodic_elements not in ["ignore", "exclude", "split"]:
+            raise ValueError(
+                f"Invalid value for 'periodic_elements'. Expected one of ['include', 'exclude', 'split'] but received: {periodic_elements}"
```
```suggestion
                f"Invalid value for 'periodic_elements'. Expected one of ['exclude', 'split', 'ignore'] but received: {periodic_elements}"
```

erogluorhan reviewed (commented):
I've resolved the previous comments and now added a few more as part of my second round of review

philipc2 commented:
> I've resolved the previous comments and now added a few more as part of my second round of review

Thank you!

erogluorhan reviewed (commented):
I am now good with all of the code here. 

This PR seems to be reducing the code coverage though, and some of them  looks real (e.g. `_correct_central_longitude` or part of it does not seem to be covered). I might be missing some contextvfrom the rest of the review process though. Could you clarify?

philipc2 commented:
> I am now good with all of the code here.
> 
> This PR seems to be reducing the code coverage though, and some of them looks real (e.g. `_correct_central_longitude` or part of it does not seem to be covered). I might be missing some contextvfrom the rest of the review process though. Could you clarify?

Thanks for the comprehensive review! I've made a few changes since writing the initial tests as I've been addressing the reviews. I'll get those added and look into improving the code coverage.

erogluorhan reviewed (approved):
This is great work and marks a visualization milestone with the projections support it brings!
